packages/mattermost-redux/src/types/actions.ts

```diff
@@ -1,30 +1,32 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
+import {ThunkAction} from 'redux-thunk';
 import {AnyAction} from 'redux';
 import {BatchAction} from 'redux-batched-actions';
 
 import {GlobalState} from '@mattermost/types/store';
+import {ServerError} from '@mattermost/types/errors';
 
 export type GetStateFunc = () => GlobalState;
 export type GenericAction = AnyAction;
 export type Thunk = (b: DispatchFunc, a: GetStateFunc) => Promise<ActionResult> | ActionResult;
 
 export type Action = GenericAction | Thunk | BatchAction | ActionFunc;
 
-export type ActionResult<Data = any, Error = any> = {
+export type ActionResult<Data = unknown, Error = unknown> = {
     data?: Data;
     error?: Error;
 };
 
-export type DispatchFunc = (action: Action, getState?: GetStateFunc | null) => Promise<ActionResult>;
+export type DispatchFunc = (
+    action: Action,
+    getState?: GetStateFunc | null
+) => Promise<ActionResult>;
 
-/**
- * Return type of a redux action.
- * @usage
- * ActionFunc<ReturnTypeOfData, ErrorType>
- */
-export type ActionFunc<Data = any, Error = any> = (
-    dispatch: DispatchFunc,
-    getState: GetStateFunc
-) => Promise<ActionResult<Data, Error> | Array<ActionResult<Data, Error>>> | ActionResult<Data, Error>;
+export type ActionFunc<Data = any, State = GlobalState, Error = ServerError, ExtraThunkArg = null> = ThunkAction<
```

Comment:
Similarly, this is sort of what I got to for this one. It also sort of replaces `Action` from above
```ts
export type AppAction<ReturnType, State extends GlobalState = GlobalState> =
    ThunkAction<ReturnType, State, never, AnyAction> |
    AnyAction;
```
I think I didn't include `ActionResult` or the `Promise` part here because I was trying to stop having every thunk action needing to return `{data: true}` when successful

Reply:
Also, I don't think we need to parameterize `ExtraThunkArg` here since we won't be overwriting that anywhere because its defined when the store is initialized. We can just pass that as `never` into `ThunkAction`.

I do have some ideas for how we can use it (like by passing a `Client4` instance instead of relying on the global singleton), but that would probably be shared across the whole code base